File: phoenix-ios/phoenix-ios/IosTCPSocket.swift
import Foundation
import Network
import PhoenixShared

class IosTCPSocket: TCPSocket {

    let connection: NWConnection

    init(_ connection: NWConnection) {
        self.connection = connection
    }

    func close() {
        connection.cancel()
    }

    func receive(min: Int32, max: Int32, completionHandler: @escaping (TCPSocketResult<KotlinByteArray>?, Error?) -> ()) {
        connection.receive(minimumIncompleteLength: Int(min), maximumLength: Int(max)) { data, context, isComplete, error in
            if let error = error {
                completionHandler(TCPSocketResultFailure(error: error.toIOException()), nil)
            } else if let data = data {
                let bytes = DarwinUtilsKt.toByteArray(data)
                completionHandler(TCPSocketResultSuccess(result: bytes), nil)
            } else if isComplete {
                completionHandler(TCPSocketResultFailure(error: TCPSocketIOException.ConnectionClosed()), nil)
            } else {
                completionHandler(TCPSocketResultFailure(error: TCPSocketIOException.Unknown(message: nil)), nil)
            }
        }
    }

    func send(bytes: KotlinByteArray?, flush: Bool, completionHandler: @escaping (TCPSocketResult<KotlinUnit>?, Error?) -> ()) {
        let pinned = bytes?.kotlinPin()
        let data = pinned?.toData()
        connection.send(content: data, isComplete: flush, completion: .contentProcessed { error in
            pinned?.unpin()
            completionHandler(error.map { TCPSocketResultFailure(error: $0.toIOException()) } ?? TCPSocketResultSuccess(result: KotlinUnit()), nil)
        })
    }

    class Builder : TCPSocketBuilder {
        func connect(host: String, port: Int32, completionHandler: @escaping (TCPSocketResult<TCPSocket>?, Error?) -> Void) {
            let connection = NWConnection(host: .init(host), port: .init(integerLiteral: UInt16(port)), using: .tcp)
            connection.stateUpdateHandler = {
                switch $0 {
                case .ready:
                    completionHandler(TCPSocketResultSuccess(result: IosTCPSocket(connection)), nil)
                    connection.stateUpdateHandler = nil
                case .failed(let error), .waiting(let error):
                    completionHandler(TCPSocketResultFailure(error: error.toIOException()), nil)
                    connection.stateUpdateHandler = nil
                default: break
                }
            }
            connection.start(queue: .main)
        }
    }

}

extension NWError {
    func toIOException() -> TCPSocketIOException {
        switch self {
        case .posix(let code):
            switch code {
            case POSIXError.ECONNREFUSED:
                return TCPSocketIOException.ConnectionRefused()
            case POSIXError.ECONNRESET:
                return TCPSocketIOException.ConnectionClosed()
            default:
                return TCPSocketIOException.Unknown(message: debugDescription)
            }
        default:
            return TCPSocketIOException.Unknown(message: debugDescription)
        }
    }
}

File: phoenix-shared/src/iosMain/kotlin/fr/acinq/phoenix/io/IosTcpSocket.kt
package fr.acinq.phoenix.io

import fr.acinq.phoenix.utils.AtomicOnce
import kotlinx.cinterop.addressOf
import kotlinx.cinterop.pin
import kotlinx.cinterop.usePinned
import platform.Foundation.NSData
import platform.Foundation.dataWithBytesNoCopy
import platform.posix.memcpy
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

@OptIn(ExperimentalUnsignedTypes::class)
class IosTcpSocket(val connection: ConnectionBridge) : TcpSocket {
    interface ConnectionBridge {
        fun receive(minimumIncompleteLength: Int, maximumLength: Int, completion: (NSData?, Boolean, TcpSocket.IOException?) -> Unit)
        fun send(content: NSData?, isComplete: Boolean, completion: (TcpSocket.IOException?) -> Unit)
        fun cancel()

        interface Builder {
            fun connect(host: String, port: Int, tls: Boolean, completion: (ConnectionBridge?, TcpSocket.IOException?) -> Unit)

            companion object {
                var native: Builder by AtomicOnce()
            }
        }
    }

    override suspend fun send(bytes: ByteArray?, flush: Boolean): Unit =
        suspendCoroutine { continuation ->
            val pinned = bytes?.pin()
            val data = pinned?.let { NSData.dataWithBytesNoCopy(it.addressOf(0), bytes.size.toULong(), false) }
            connection.send(data, flush) { error ->
                pinned?.unpin()
                if (error != null) continuation.resumeWithException(error)
                else continuation.resume(Unit)
            }
        }

    override suspend fun receiveFully(buffer: ByteArray) {
        receive(buffer.size, buffer)
    }

    override suspend fun receiveAvailable(buffer: ByteArray): Int =
        receive(0, buffer)

    private suspend fun receive(min: Int, buffer: ByteArray): Int =
        suspendCoroutine { continuation ->
            connection.receive(min, buffer.size) { data, isComplete, error ->
                when {
                    error != null -> continuation.resumeWithException(error)
                    data != null -> {
                        buffer.usePinned { pinned ->
                            memcpy(pinned.addressOf(0), data.bytes, data.length)
                        }
                        continuation.resume(data.length.toInt())
                    }
                    isComplete -> continuation.resumeWithException(TcpSocket.IOException.ConnectionClosed)
                    else -> error("NWConnection receive returned no data, no error, and no completion")
                }
            }
        }

    override fun close() {
        connection.cancel()
    }
}

internal actual object PlatformSocketBuilder : TcpSocket.Builder {
    override suspend fun connect(host: String, port: Int, tls: Boolean): TcpSocket =
        suspendCoroutine { continuation ->
            IosTcpSocket.ConnectionBridge.Builder.native.connect(host, port, tls) { connection, error ->
                when {
                    error != null -> continuation.resumeWithException(error)
                    connection != null -> continuation.resume(IosTcpSocket(connection))
                }
            }
        }
}
